New Adapter: TPMN

tpmn-admin opened the pull request:
update doc : https://github.com/prebid/prebid.github.io/pull/4736


Dear Prebid Team! 

Please review the changes and let us know if there's anything further to set. Once we get confirmation this has been merged we will update our adapter documentation as well as requested.

onkarvhanumante commented on adapters/tpmn/tpmn.go:
```diff
@@ -0,0 +1,159 @@
+	return bidResponse, nil
+}
+
+func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Banner != nil {
+				return openrtb_ext.BidTypeBanner, nil
+			} else if imp.Video != nil {
+				return openrtb_ext.BidTypeVideo, nil
+			} else if imp.Native != nil {
+				return openrtb_ext.BidTypeNative, nil
+			}
+		}
+	}
+
+	return "", &errortypes.BadInput{
+		Message: fmt.Sprintf("Failed to find a supported media type impression \"%s\"", impID),
+	}
```
Warning: this is an anti-pattern. This code assumes that if there is multi-format request then mediatype is defaulting to banner.
Ideally, Prebid server expects media type set in adapter response. we recommend implementing a pattern where adapter server sets the [MType](https://github.com/prebid/openrtb/blob/53ac1b9ca7ddf860988ff69bf5018b32ced1fbc5/openrtb2/bid.go#L334) field in the response to accurately determine the media type for the impression.

tpmn-admin replied:
replace MakeBids responseStatus code check logic :
[8cec466](https://github.com/prebid/prebid-server/pull/2947/commits/8cec4663d47621ff2654b38692c5391800b7b3de)

onkarvhanumante commented on static/bidder-info/tpmn.yaml:
```diff
@@ -0,0 +1,20 @@
+endpoint: "https://gat.tpmn.io/ortb/pbs_bidder"
```
endpoint is reachable

```
oh0387@oh0387s-MacBook-Pro prebid-server % curl -i https://gat.tpmn.io/ortb/pbs_bidder
HTTP/2 204
date: Thu, 20 Jul 2023 06:13:36 GMT
via: 1.1 google
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
```

onkarvhanumante commented on static/bidder-info/tpmn.yaml:
```diff
@@ -0,0 +1,20 @@
+    mediaTypes:
+      - banner
+      - video
+      - native
```
should add json test for native imps

onkarvhanumante commented on static/bidder-info/tpmn.yaml:
```diff
@@ -0,0 +1,20 @@
+  email: "prebid@tpmn.io"
+modifyingVastXmlAllowed: true
+endpointCompression: GZIP
+capabilities:
+  app:
+    mediaTypes:
+      - banner
+      - video
+      - native
+  site:
+    mediaTypes:
+      - banner
+      - video
+      - native
```
json tests doesn't cover all these paths. please add missing tests for all supported capabilities

tpmn-admin replied:
update commit :

[1dce2d4](https://github.com/prebid/prebid-server/pull/2947/commits/1dce2d4789e2cd61f91fb3ef5aa606d2713c5289)

[15dc048](https://github.com/prebid/prebid-server/pull/2947/commits/15dc0487250ed476324b4512938bdea052cfb3d8)

onkarvhanumante commented on static/bidder-info/tpmn.yaml:
```diff
@@ -0,0 +1,20 @@
+      - banner
+      - video
+      - native
+userSync:
+  redirect:
+    url: "https://gat.tpmn.io/sync/redirect?gdpr={{.GDPR}}&gdpr_consent={{.GDPRConsent}}&us_privacy={{.USPrivacy}}&redir={{.RedirectURL}}"
+    support-cors: false
```
`userMacro` is missing

tpmn-admin replied:
update Commit : 
[1c09ea0](https://github.com/prebid/prebid-server/pull/2947/commits/1c09ea0990396b38b7baa43627bf8761dc90acea)

onkarvhanumante commented on openrtb_ext/imp_tpmn.go:
```diff
@@ -0,0 +1,6 @@
+package openrtb_ext
+
+// ExtImpTpmn defines adman specifiec param
```
nit pick: replace `adman` with your adapter name

tpmn-admin replied:
update commit : 
[3c38636](https://github.com/prebid/prebid-server/pull/2947/commits/3c38636a36aee5a3fe0827001cfcd40c4fa96416)

onkarvhanumante commented on adapters/tpmn/tpmn.go:
```diff
@@ -0,0 +1,159 @@
+	case http.StatusNoContent:
+		return nil, nil
+	case http.StatusBadRequest:
+		return nil, []error{errors.New("bad request from publisher")}
+	case http.StatusOK:
+		break
+	default:
+		return nil, []error{fmt.Errorf("unexpected response status code: %v", responseData.StatusCode)}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{fmt.Errorf("bid response unmarshal: %v", err)}
+	}
+
```
https://github.com/prebid/prebid-server/blob/eb338b6c4c68b778cf103bb7028dd1245ab1aee9/adapters/bidder.go#L40-L41

please use `errortypes.BadInput` and `errortypes.BadServerResponse ` as mentioned above

tpmn-admin replied:
update commit : 
[8cec466](https://github.com/prebid/prebid-server/pull/2947/commits/8cec4663d47621ff2654b38692c5391800b7b3de)

onkarvhanumante commented on adapters/tpmn/tpmn.go:
```diff
@@ -0,0 +1,159 @@
+	for _, seatBid := range response.SeatBid {
+		for i := range seatBid.Bid {
+			bidType, err := getMediaTypeForImp(seatBid.Bid[i].ImpID, request.Imp)
+			if err != nil {
+				return nil, []error{err}
+			}
```
might consider collecting error and continuing for remaining valid seat bids

```
	var errs []error
	...

	for _, seatBid := range response.SeatBid {
		for i := range seatBid.Bid {
			bidType, err := getMediaTypeForImp(seatBid.Bid[i].ImpID, request.Imp)
			if err != nil {
				errs = append(errs, err)
				continue
			}
			b := &adapters.TypedBid{
				Bid:     &seatBid.Bid[i],
				BidType: bidType,
			}
			bidResponse.Bids = append(bidResponse.Bids, b)
		}
	}
```

tpmn-admin replied:
update Commit : 

[c2d5827](https://github.com/prebid/prebid-server/pull/2947/commits/c2d5827ea77f7a2b5386a340f410bb37fd8bbdf8)

tpmn-admin replied:
Thank you for the valuable feedback. The default type of getMediaTypeForImp has been changed to 'banner', which allows it to handle the error part without consideration.

tpmn-admin replied:
Relevant commit : [8cec466](https://github.com/prebid/prebid-server/pull/2947/commits/8cec4663d47621ff2654b38692c5391800b7b3de)

onkarvhanumante commented on adapters/tpmn/tpmn.go:
```diff
@@ -0,0 +1,159 @@
+
+// TpmnAdapter struct
+type adapter struct {
+	URI string
```
`adapter` is private member of `tpmn` package. `URI` can also be private
```
type adapter struct {
	uri string
}
```

tpmn-admin replied:
update commit : 
[458faa1](https://github.com/prebid/prebid-server/pull/2947/commits/458faa1a7c7943115695ce8d227f8abaa283d7fe)

onkarvhanumante commented on adapters/tpmn/tpmn_test.go:
```diff
@@ -0,0 +1,20 @@
+
+func TestJsonSamples(t *testing.T) {
+	bidder, buildErr := Builder(openrtb_ext.BidderTpmn, config.Adapter{
+		Endpoint: "http://example.com/pserver"}, config.Server{ExternalUrl: "http://hosturl.com", GvlID: 1, DataCenter: "2"})
```
This is correct. But suggesting to match `Endpoint` url with `https://gat.tpmn.io/ortb/pbs_bidder` mentioned in bidder params

tpmn-admin replied:
update commit : 
[1390874](https://github.com/prebid/prebid-server/pull/2947/commits/1390874130016765f0134fe3a2321440c995f2dd)
[7896970](https://github.com/prebid/prebid-server/pull/2947/commits/7896970cb691e5e411608c3e012f1ffd32639757)

onkarvhanumante commented on openrtb_ext/imp_tpmn.go:
```diff
@@ -0,0 +1,6 @@
+
+// ExtImpTpmn defines adman specifiec param
+type ExtImpTpmn struct {
+	inventoryId int `json:"inventoryId"`
```
`ExtImpTpmn` will be used with `unmarshal` in adapter code. Note that `unmarshal` works with  public members of struct

```
type ExtImpTpmn struct {
	InventoryId int `json:"inventoryId"`
}
```

tpmn-admin replied:
Thank you for your kind explanation.

update commit : 
[3c38636](https://github.com/prebid/prebid-server/pull/2947/commits/3c38636a36aee5a3fe0827001cfcd40c4fa96416)

tpmn-admin commented:
Dear Prebid Team! 
We have incorporated the changes based on the feedback you provided. 
The modified code has been documented with the link to the commit. 
Additionally, all the addressed feedback has been marked as 'Resolved'.

Thx

gargcreation1992 commented on adapters/tpmn/params_test.go:
```diff
@@ -0,0 +1,45 @@
+
+var validParams = []string{
+	`{"inventoryId": 10000001}`,
+}
```
nitpick - Generally, we read code from top to bottom. Declaring variable after it has already been used is not generally considered a good coding practice. Requesting you to please move this piece of code to the top(after imports statement) of params_test.go file .

tpmn-admin replied:
You are right. 
Since I'm not familiar with Go language, I referenced a lot of code from other companies and structured my code similarly. 
Many partners placed variables at the bottom of the code. 
That part also seems to need checking.

tpmn-admin replied:
update commit :
[1f22b27](https://github.com/prebid/prebid-server/pull/2947/commits/1f22b27a0bc801cd54d52dc7a9f1b9ffe8223894)

gargcreation1992 commented on adapters/tpmn/tpmn.go:
```diff
@@ -0,0 +1,153 @@
+		return &errortypes.BadInput{
+			Message: "Wrong TPMN bidder ext: " + err.Error(),
+		}
+	}
```
I do not think you need this check here. This check is not doing anything but unmarshalling json to `openrtb_ext.ExtImpTpmn` type that is unnecessary check. Requesting you to please remove it.

tpmn-admin replied:
update commit :
[f861d08](https://github.com/prebid/prebid-server/pull/2947/commits/f861d080a269655a4ac44d57daa2a10b7bb0256d)

gargcreation1992 commented on adapters/tpmn/tpmn.go:
```diff
@@ -0,0 +1,153 @@
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Invalid Status Returned: %d. Run with request.debug = 1 for more info", responseData.StatusCode),
+		}}
+	}
```
This is correct. But you could also use below in-built checks - 
```
if adapters.IsResponseStatusCodeNoContent(response) {
		return nil, nil
	}

	if err := adapters.CheckResponseStatusCodeForErrors(response); err != nil {
		return nil, []error{err}
	}
	```

tpmn-admin replied:
update commit :
[ba8be5f](https://github.com/prebid/prebid-server/pull/2947/commits/ba8be5fca27cb98322e09f0aed51d2491435b58d)

gargcreation1992 commented on adapters/tpmn/tpmn.go:
```diff
@@ -0,0 +1,153 @@
+			}
+			break
+		}
+	}
```
Warning - This is an anti-pattern. This code assumes that for multi-format bid request, mediatype will be set to type banner by default. I would recommend to set the mediaType based on bidder response. Your bidder should be able to set openrtb2.Bid.MType field and use that field here to get mediaTypeForImp. Even if your bidder does not support multi-format request, I would highly encourage you to make this change.

tpmn-admin replied:
update commit : [e35524c](https://github.com/prebid/prebid-server/pull/2947/commits/e35524c2a89ce402e50f7da21400b9d2e301c6cb)

gargcreation1992 commented on static/bidder-info/tpmn.yaml:
```diff
@@ -0,0 +1,20 @@
+      - native
+userSync:
+  redirect:
+    url: "https://gat.tpmn.io/sync/redirect?gdpr={{.GDPR}}&gdpr_consent={{.GDPRConsent}}&us_privacy={{.USPrivacy}}&redir={{.RedirectURL}}"
```
The user sync URL is not setting the UID. Could you please check? Trying the following URL - 
```
https://gat.tpmn.io/sync/redirect?gdpr=&gdpr_consent=&us_privacy=&redir=http%3A%2F%2Flocalhost%3A8000%2Fsetuid%3Fbidder%3Dtpmn%26gdpr%3D%26gdpr_consent%3D%26gpp%3D%26gpp_sid%3D%26f%3Di%26uid%3D%24UID
```

tpmn-admin replied:
The macro function did not work normally, so it was fixed.
Now you can check that it works normally.

Thx!

gargcreation1992 replied:
It is working now. Thanks!

tpmn-admin replied:
Good! Thank you for checking.

tpmn-admin commented:
Dear Prebid Team!
I have completed the updates regarding the additional feedback. 
For each feedback, I have linked the commit address for easy reference.

gargcreation1992 reviewed (commented):
Just one more small change, g2g after that!

tpmn-admin commented on adapters/tpmn/tpmn.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, _ *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		// no bid response
+		return nil, nil
```
Opps!  here was some code that needed to be removed, and I had missed that part. 
Your feedback was much appreciated.

tpmn-admin commented:
> Just one more small change, g2g after that!

fix it.

onkarvhanumante commented on static/bidder-info/tpmn.yaml:
```diff
@@ -0,0 +1,20 @@
+endpoint: "https://gat.tpmn.io/ortb/pbs_bidder"
+maintainer:
+  email: "prebid@tpmn.io"
```
@tpmn-admin , Prebid team has sent verification email to confirm accuracy of above email. Requesting to respond back on email with "received" message

tpmn-admin replied:
received

onkarvhanumante replied:
@tpmn-admin  please respond back on email

tpmn-admin replied:
@onkarvhanumante sorry for late. I thought I would reply to the mail, but it didn't get sent. I replied to your email a few minutes ago.

gargcreation1992 commented on adapters/tpmn/tpmn.go:
```diff
@@ -0,0 +1,143 @@
+			continue
+		}
+
+		if err := preprocessExtensions(&imp); err != nil {
+			errs = append(errs, err)
+			continue
+		}
```
While taking a final look, this check seems unnecessary. Usually, you would like to process the extension if you want to use the fields from `adapters.ExtImpBidder` in your adapter code and that is clearly not the case here. The code is just doing json unmarshal. I would recommend to remove this check. thanks!

tpmn-admin replied:
I agree with what you said. After receiving and checking the message, I found out that the prebid server already does the check for the presence of the common extension filed by default. it is code unnecessary. I have removed the code as requested. Additionally, the 'no-imp-ext.json' test file has also been modified.

update commit : [3f5af95](https://github.com/prebid/prebid-server/pull/2947/commits/3f5af95e6f53af9f0df0f8aaa84b2cd94e644362)

gargcreation1992 reviewed (commented):
Thanks for being patient!
While taking a final look, I came across one unnecessary check. Requesting you to please remove it.
